File: Data Structures & Algorithms/implement-stack-using-queues/submission-0.py
class MyStack:
    q1 = None
    def __init__(self):
      self.q1 = deque()  

    def push(self, x: int) -> None:
        self.q1.append(x)

    def pop(self) -> int:
        q2 = deque()
        top = 0
        while not self.empty():
            x = self.q1.popleft()
            if self.empty():
                top = x
                break
            q2.append(x)
        self.q1 = q2
        return top
    def top(self) -> int:
        q2 = deque()
        top = 0
        print("top",self.empty())
        while not self.empty():
            x = self.q1.popleft()
            if self.empty():
                top = x
            q2.append(x)
        self.q1 = q2
        return top
    def empty(self) -> bool:
        return True if len(self.q1) == 0 else False


# Your MyStack object will be instantiated and called as such:
# obj = MyStack()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.top()
# param_4 = obj.empty()

File: Data Structures & Algorithms/range-sum-query-2d-immutable/submission-0.cpp
class NumMatrix {
private:
    vector<vector<int>> sumMat;
public:
    NumMatrix(vector<vector<int>>& matrix) {
        int n = matrix.size(), m = matrix[0].size();
        sumMat = vector<vector<int>>(n+1, vector<int>(m+1,0));
        for(int i=0; i<n; i++){
            int prefix = 0;
            for(int j=0; j<m; j++){
                prefix += matrix[i][j];
                sumMat[i+1][j+1] = sumMat[i][j+1] + prefix;
            }
        }
        // for(int i=0; i<=n; i++){
        //     for(int j=0; j<=m; j++){
        //         cout<<sumMat[i][j]<<" ";
        //     }
        //     cout<<endl;
        // }
    }
    
    int sumRegion(int row1, int col1, int row2, int col2) {
        int last = sumMat[row2+1][col2+1];
        int top = sumMat[row1][col1];
        int x1 = sumMat[row1][col2+1];
        int x2 = sumMat[row2+1][col1];
        return top+last-x1-x2;
    }
};

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix* obj = new NumMatrix(matrix);
 * int param_1 = obj->sumRegion(row1,col1,row2,col2);
 */